### PR TITLE
refactor `tool_doc!`

### PR DESCRIPTION
**Pull request by onur-ozkan**
resolves https://github.com/rust-lang/rust/pull/112211#discussion_r1215190510

### CODE REVIEW

**Comment by albertlarsan68** [comment]
Thanks for the PR!
@bors r+

**Review by eval-exec — APPROVED** [review]

**Comment by ozkanonur** [comment]
```rs
failures:

---- [run-make] tests\run-make\rustdoc-scrape-examples-multiple stdout ----

error: make failed
status: exit code: 2
command: "make"
--- stdout -------------------------------
make[1]: Entering directory '/c/a/rust/rust/tests/run-make/rustdoc-scrape-examples-multiple'
PATH="/c/a/rust/rust/build/x86_64-pc-windows-gnu/test/run-make/rustdoc-scrape-examples-multiple/rustdoc-scrape-examples-multiple:C:\a\rust\rust\build\x86_64-pc-windows-gnu\stage2\bin:/c/a/rust/rust/build/x86_64-pc-windows-gnu/stage0-bootstrap-tools/x86_64-pc-windows-gnu/release/deps:/c/a/rust/rust/build/x86_64-pc-windows-gnu/stage0/bin:/c/a/rust/rust/ninja:/c/a/rust/rust/mingw64/bin:/c/hostedtoolcache/windows/Python/3.11.3/x64/Scripts:/c/hostedtoolcache/windows/Python/3.11.3/x64:/usr/bin:/c/a/rust/rust/sccache:/c/Program Files/MongoDB/Server/5.0/bin:/c/aliyun-cli:/c/vcpkg:/c/cf-cli:/c/Program Files (x86)/NSIS:/c/tools/zstd:/c/Program Files/Mercurial:/c/hostedtoolcache/windows/stack/2.9.3/x64:/c/cabal/bin:/c/ghcup/bin:/c/Program Files/dotnet:/c/mysql/bin:/c/Program Files/R/R-4.3.0/bin/x64:/c/SeleniumWebDrivers/GeckoDriver:/c/Program Files (x86)/sbt/bin:/c/Program Files (x86)/GitHub CLI:/c/Program Files/Git/bin:/c/Program Files (x86)/pipx_bin:/c/npm/prefix:/c/hostedtoolcache/windows/go/1.20.3/x64/bin:/c/hostedtoolcache/windows/Python/3.7.9/x64/Scripts:/c/hostedtoolcache/windows/Python/3.7.9/x64:/c/hostedtoolcache/windows/Ruby/2.5.9/x64/bin:/c/Program Files/OpenSSL/bin:/c/tools/kotlinc/bin:/c/hostedtoolcache/windows/Java_Temurin-Hotspot_jdk/8.0.372-7/x64/bin:/c/Program Files/ImageMagick-7.1.1-Q16-HDRI:/c/Program Files (x86)/Microsoft SDKs/Azure/CLI2/wbin:/c/ProgramData/kind:/c/Program Files/Eclipse Foundation/jdk-8.0.302.8-hotspot/bin:/c/Windows/system32:/c/Windows:/c/Windows/System32/Wbem:/c/Windows/System32/WindowsPowerShell/v1.0:/c/Windows/System32/OpenSSH:/c/ProgramData/Chocolatey/bin:/c/Program Files/PowerShell/7:/c/Program Files/Microsoft/Web Platform Installer:/c/Program Files/Microsoft SQL Server/130/Tools/Binn:/c/Program Files/Microsoft SQL Server/Client SDK/ODBC/170/Tools/Binn:/c/Program Files (x86)/Windows Kits/10/Windows Performance Toolkit:/c/Program Files (x86)/Microsoft SQL Server/110/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/120/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/130/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/140/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/150/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/160/DTS/Binn:/c/Strawberry/c/bin:/c/Strawberry/perl/site/bin:/c/Strawberry/perl/bin:/c/ProgramData/chocolatey/lib/pulumi/tools/Pulumi/bin:/c/Program Files/TortoiseSVN/bin:/c/Program Files/CMake/bin:/c/ProgramData/chocolatey/lib/maven/apache-maven-3.8.7/bin:/c/Program Files/Microsoft Service Fabric/bin/Fabric/Fabric.Code:/c/Program Files/Microsoft SDKs/Service Fabric/Tools/ServiceFabricLocalClusterManager:/c/Program Files/nodejs:/c/Program Files/Git/cmd:/c/Program Files/Git/mingw64/bin:/c/Program Files/Git/usr/bin:/c/Program Files/GitHub CLI:/c/tools/php:/c/Program Files (x86)/sbt/bin:/c/SeleniumWebDrivers/ChromeDriver:/c/SeleniumWebDrivers/EdgeDriver:/c/Program Files/Amazon/AWSCLIV2:/c/Program Files/Amazon/SessionManagerPlugin/bin:/c/Program Files/Amazon/AWSSAMCLI/bin:/c/Program Files (x86)/Google/Cloud SDK/google-cloud-sdk/bin:/c/Program Files (x86)/Microsoft BizTalk Server:/c/Program Files/LLVM/bin:/c/Users/runneradmin/.dotnet/tools:/c/Users/runneradmin/.cargo/bin:/c/Users/runneradmin/AppData/Local/Microsoft/WindowsApps" 'C:\a\rust\rust\build\x86_64-pc-windows-gnu\stage2\bin\rustc.exe' --out-dir /c/a/rust/rust/build/x86_64-pc-windows-gnu/test/run-make/rustdoc-scrape-examples-multiple/rustdoc-scrape-examples-multiple -L /c/a/rust/rust/build/x86_64-pc-windows-gnu/test/run-make/rustdoc-scrape-examples-multiple/rustdoc-scrape-examples-multiple  src/lib.rs --crate-name foobar --crate-type lib --emit=metadata
PATH="/c/a/rust/rust/build/x86_64-pc-windows-gnu/test/run-make/rustdoc-scrape-examples-multiple/rustdoc-scrape-examples-multiple:C:\a\rust\rust\build\x86_64-pc-windows-gnu\stage2\bin:/c/a/rust/rust/build/x86_64-pc-windows-gnu/stage0-bootstrap-tools/x86_64-pc-windows-gnu/release/deps:/c/a/rust/rust/build/x86_64-pc-windows-gnu/stage0/bin:/c/a/rust/rust/ninja:/c/a/rust/rust/mingw64/bin:/c/hostedtoolcache/windows/Python/3.11.3/x64/Scripts:/c/hostedtoolcache/windows/Python/3.11.3/x64:/usr/bin:/c/a/rust/rust/sccache:/c/Program Files/MongoDB/Server/5.0/bin:/c/aliyun-cli:/c/vcpkg:/c/cf-cli:/c/Program Files (x86)/NSIS:/c/tools/zstd:/c/Program Files/Mercurial:/c/hostedtoolcache/windows/stack/2.9.3/x64:/c/cabal/bin:/c/ghcup/bin:/c/Program Files/dotnet:/c/mysql/bin:/c/Program Files/R/R-4.3.0/bin/x64:/c/SeleniumWebDrivers/GeckoDriver:/c/Program Files (x86)/sbt/bin:/c/Program Files (x86)/GitHub CLI:/c/Program Files/Git/bin:/c/Program Files (x86)/pipx_bin:/c/npm/prefix:/c/hostedtoolcache/windows/go/1.20.3/x64/bin:/c/hostedtoolcache/windows/Python/3.7.9/x64/Scripts:/c/hostedtoolcache/windows/Python/3.7.9/x64:/c/hostedtoolcache/windows/Ruby/2.5.9/x64/bin:/c/Program Files/OpenSSL/bin:/c/tools/kotlinc/bin:/c/hostedtoolcache/windows/Java_Temurin-Hotspot_jdk/8.0.372-7/x64/bin:/c/Program Files/ImageMagick-7.1.1-Q16-HDRI:/c/Program Files (x86)/Microsoft SDKs/Azure/CLI2/wbin:/c/ProgramData/kind:/c/Program Files/Eclipse Foundation/jdk-8.0.302.8-hotspot/bin:/c/Windows/system32:/c/Windows:/c/Windows/System32/Wbem:/c/Windows/System32/WindowsPowerShell/v1.0:/c/Windows/System32/OpenSSH:/c/ProgramData/Chocolatey/bin:/c/Program Files/PowerShell/7:/c/Program Files/Microsoft/Web Platform Installer:/c/Program Files/Microsoft SQL Server/130/Tools/Binn:/c/Program Files/Microsoft SQL Server/Client SDK/ODBC/170/Tools/Binn:/c/Program Files (x86)/Windows Kits/10/Windows Performance Toolkit:/c/Program Files (x86)/Microsoft SQL Server/110/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/120/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/130/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/140/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/150/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/160/DTS/Binn:/c/Strawberry/c/bin:/c/Strawberry/perl/site/bin:/c/Strawberry/perl/bin:/c/ProgramData/chocolatey/lib/pulumi/tools/Pulumi/bin:/c/Program Files/TortoiseSVN/bin:/c/Program Files/CMake/bin:/c/ProgramData/chocolatey/lib/maven/apache-maven-3.8.7/bin:/c/Program Files/Microsoft Service Fabric/bin/Fabric/Fabric.Code:/c/Program Files/Microsoft SDKs/Service Fabric/Tools/ServiceFabricLocalClusterManager:/c/Program Files/nodejs:/c/Program Files/Git/cmd:/c/Program Files/Git/mingw64/bin:/c/Program Files/Git/usr/bin:/c/Program Files/GitHub CLI:/c/tools/php:/c/Program Files (x86)/sbt/bin:/c/SeleniumWebDrivers/ChromeDriver:/c/SeleniumWebDrivers/EdgeDriver:/c/Program Files/Amazon/AWSCLIV2:/c/Program Files/Amazon/SessionManagerPlugin/bin:/c/Program Files/Amazon/AWSSAMCLI/bin:/c/Program Files (x86)/Google/Cloud SDK/google-cloud-sdk/bin:/c/Program Files (x86)/Microsoft BizTalk Server:/c/Program Files/LLVM/bin:/c/Users/runneradmin/.dotnet/tools:/c/Users/runneradmin/.cargo/bin:/c/Users/runneradmin/AppData/Local/Microsoft/WindowsApps" 'C:\a\rust\rust\build\x86_64-pc-windows-gnu\stage2\bin\rustdoc.exe' -L C:\a\rust\rust\build\x86_64-pc-windows-gnu\stage2\lib\rustlib\x86_64-pc-windows-gnu\lib examples/ex.rs --crate-name ex --crate-type bin --output "/c/a/rust/rust/build/x86_64-pc-windows-gnu/test/run-make/rustdoc-scrape-examples-multiple/rustdoc-scrape-examples-multiple/rustdoc" \
  --extern foobar=/c/a/rust/rust/build/x86_64-pc-windows-gnu/test/run-make/rustdoc-scrape-examples-multiple/rustdoc-scrape-examples-multiple/libfoobar.rmeta \
	-Z unstable-options \
	--scrape-examples-output-path /c/a/rust/rust/build/x86_64-pc-windows-gnu/test/run-make/rustdoc-scrape-examples-multiple/rustdoc-scrape-examples-multiple/ex.calls \
	--scrape-examples-target-crate foobar \
	
make[1]: Leaving directory '/c/a/rust/rust/tests/run-make/rustdoc-scrape-examples-multiple'
------------------------------------------
--- stderr -------------------------------
thread 'main' has overflowed its stack
make[1]: *** [scrape.mk:6: /c/a/rust/rust/build/x86_64-pc-windows-gnu/test/run-make/rustdoc-scrape-examples-multiple/rustdoc-scrape-examples-multiple/ex.calls] Error 12
```

Hmm. Suspicious :thinking:

@bors retry

**Comment by matthiaskrgr** [comment]
@bors r-
also stack overflow'd in the rollup   https://github.com/rust-lang/rust/pull/112264 :bomb:

**Comment by ozkanonur** [comment]
I suspected that adding an additional argument to the `tool_doc!` macro was causing stack overflows(probably because of the recursion) randomly in some of our tests. I decided to make `$krate` more generic which is now `$extra_args` to allow us to include `--lib`(or any other argument if needed) within it as well.

It would be nice if you can make another review @albertlarsan68 and also I would like to know your(@jyn514) opinion on this change :)

**Review by albertlarsan68 — APPROVED** [review]
LGTM, but would also like jyn's approval.
r=me if OK

File: src/bootstrap/doc.rs [hunk 1]
```diff
@@ -834,9 +842,10 @@ macro_rules! tool_doc {
                 cargo.arg("-Zskip-rustdoc-fingerprint");
                 // Only include compiler crates, no dependencies of those, such as `libc`.
                 cargo.arg("--no-deps");
-                cargo.arg("--lib");
+
                 $(
-                    cargo.arg("-p").arg($krate);
+                    let args: Vec<&str> = $extra_arg.split_whitespace().collect();
```
Comment (jyn514) [inline src/bootstrap/doc.rs:847]: there's no point in making this a Vec if you're going to call split whitespace - the interface should either be `["-p", "rustdoc", "-p", "rustdoc-json-types"]` or `"-p rustdoc -p rustdoc-json-types`. slight preference towards the latter but i don't feel strongly.

**Comment by jyn514** [comment]
yeah this seems fine :)

**Pull request merged**